TESTS/mbed_hal/flash/functional_tests/main.cpp

```diff
@@ -0,0 +1,129 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !DEVICE_FLASH
+    #error [NOT_SUPPORTED] Flash API not supported for this target
+#endif
+
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+
+#include "mbed.h"
+#include "flash_api.h"
+#include "flash_data.h"
+
+using namespace utest::v1;
+
+/*
+    return values to be checked are documented at:
+        http://arm-software.github.io/CMSIS_5/Pack/html/algorithmFunc.html#Verify
+*/
+
+#ifndef ALIGN_DOWN
+#define ALIGN_DOWN(x, a) ((x)& ~((a) - 1))
+#endif
+
+flash_t test_flash;
+
+void flash_init()
```

Comment:
Nit: I was initially confused that the test function name is the same as the function that's being tested. Any chance you could add `_test` or something to the test function?